pycid/core/macid.py

```diff
@@ -56,46 +103,114 @@ def get_all_pure_ne_in_sg(
+
+        else:
+            # pure NE finder
+            all_pure_ne_in_sg: List[List[StochasticFunctionCPD]] = []
+            for pp in self.joint_pure_policies(decisions_in_sg):
 
-    def get_all_pure_spe(self) -> List[List[StochasticFunctionCPD]]:
-        """Return a list of all pure subgame perfect Nash equilbiria (SPE) in the MACIM
-        - Each SPE comes as a list of FunctionCPDs, one for each decision node in the MACID.
+                for a in agents_in_sg:  # check that each agent is happy
+                    macid.add_cpds(*pp)  # impute the policy profile
+
+                    eu_pp_agent_a = macid.expected_utility({}, agent=a)
+                    macid.add_cpds(*macid.optimal_pure_policies(agent_decs_in_sg[a])[0])
+                    max_eu_agent_a = macid.expected_utility({}, agent=a)
+                    if max_eu_agent_a > eu_pp_agent_a:  # not an NE
+                        break
+                else:
+                    # it's an NE
+                    all_pure_ne_in_sg.append(list(pp))
+
+            return all_pure_ne_in_sg
+
+    def is_nash(self, policy_profile: Iterable[StochasticFunctionCPD]) -> bool:
+        """
+        Given a policy profile, is it the case that this policy profile is a pure policy Nash equilibrium in
+        the MACID?
+        """
+        for a in self.agents:  # check that each agent is happy
+            self.add_cpds(*policy_profile)
+            eu_pp_agent_a = self.expected_utility({}, agent=a)
+            self.add_cpds(*self.optimal_pure_policies(self.agent_decisions[a])[0])
+            max_eu_agent_a = self.expected_utility({}, agent=a)
+            if max_eu_agent_a > eu_pp_agent_a:  # not an NE
+                return False
+        else:
+            # it's an NE
+            return True
+
+    def mixed_policy(
+        self,
+        agent_pure_policies: Tuple[Tuple[StochasticFunctionCPD, ...], ...],
+        prob_dist: Union[List[float], np.ndarray],
+    ) -> Iterator[StochasticFunctionCPD]:
+        """
+        Given a list of the agent's pure policies over some decisions and a distribution over these pure policies,
+        return a generator of the equivalent mixed decision rules that makes up this mixed policy
         """
-        spes: List[List[StochasticFunctionCPD]] = [[]]
 
-        # backwards induction over the sccs in the condensed relevance graph (handling tie-breaks)
-        for scc in reversed(CondensedRelevanceGraph(self).get_scc_topological_ordering()):
-            extended_spes = []
-            for partial_profile in spes:
-                self.add_cpds(*partial_profile)
-                all_ne_in_sg = self.get_all_pure_ne_in_sg(scc)
-                for ne in all_ne_in_sg:
-                    extended_spes.append(partial_profile + list(ne))
-            spes = extended_spes
-        return spes
+        num_decision_rules = len(agent_pure_policies[0])  # how many decision nodes that agent has
+
+        for i in range(num_decision_rules):
```

Comment:
why do you loop over integers i instead of just agent_pure_policies directly?

Reply:
Yeah i should change 'i' to 'dec' as this isn't clear. We have the agent's pure policies and the desired distribution over pure policies. We want a mixed behavioural policy (maybe this should also be in the method name/doc string), so we want to do this summation over dists thing for each dec node. Does that make sense?